3 week

duswnsxnxn opened the pull request:
# 3Week_김연준.md

## Title: [3Week] 김연준

### 필수 미션 요구사항 분석 & 체크리스트
- [x] 호감 표시/호감 사유 변경 후, 개별 호감 표시건에 대해서, 3시간 동안은 호감 취소와 호감 사유 변경을 할 수 없도록 작업
- [x] 네이버클라우드 플랫폼을 통한 배포(도메인 없이, IP로 접속)
### 선택 미션 요구사항 분석 & 체크리스트
- [x] 호감 표시를 받았거나, 본인에 대한 호감 사유가 변경된 경우에 알림 페이지에서 확인
- [x] 각각의 알림은 생성시에 readDate가 null 이고, 사용자가 알림을 읽으면 현재 날짜로 세팅
---

- 쿨타임 기능 구현
- 알림 기능 구현
- 네이버클라우드 플랫폼을 통한 배포
### 3주차 미션 요약

---

**[접근 방법]**



- 호감 표시, 수정 시 쿨타임
  - 호감 표시를 하면 호감 수정을 3시간 뒤에 가능하게끔 LikeablePerson에 있는 modifyUnlockDate에 쿨타임 수치 주입
  - 호감 수정을 하면 canLike부터 진입하여 쿨타임 조건문을 통해 통과되면 수정 가능
  - 조건문에는 modifyUnlockDate필드를 LocalDateTime클래스의 isAfter메소드를 이용해 현재 시간과 비교하는 로직이 있음
  - modifyUnlockDate필드를 String타입으로 호감 리스트 페이지에 출력
- 알림 기능 구현
  - 호감 표시/호감 사유 변경 이벤트 발생 시점에 알림 서비스를 끼워넣어 데이터베이스에 저장시킴
  - 호감 표시는 Like타입으로 호감 수정은 ModifyAttractiveType타입으로 저장
  - 저장된 타입을 기준으로 알림을 따로 구분하여 페이지에 출력
  - 각각의 알림은 생성시에 readDate가 null 이고, 사용자가 알림을 읽으면 현재 날짜로 세팅


**[특이사항]**

아쉬웠던 점
- 호감취소 쿨타임 처리를 깜빡하고 하지 못했습니다.
- 알림 이벤트를 따로 만들어서 처리하지 않고 InstaMemberEventListener에 알림 서비스를 끼워넣는 방식으로 해결하여 가독성이 좋지 않습니다.
- 리눅스에서 빌드 할 떄 계속 테스트 FAILED가 떠서 배포를 하지 못했습니다.

궁금했던 점
- 리눅스에서 복사한 부분의 맨 앞 글자가 잘려서 나오는 바람에 테스트 실패라고 나왔습니다.
리눅스에서 붙여넣기 할 때 원본을 유지할 수 있는 방법은 뭔가요?

**[Refactoring]**

duswnsxnxn committed:
refactor: yml리펙토링
- NotProd가 실행될 때 oauthId를 yml로부터 참조
- RSA public key is not available client side에러 해결

duswnsxnxn committed:
feat: 호감표시,수정 시 백엔드사이드 쿨타임 적용

duswnsxnxn committed:
feat: 본인에 대한 누군가의 호감 표시, 수정 알림 기능 구현

duswnsxnxn committed:
test: 알림 기능 테스트 추가 및 쿨타임 테스트 수정

duswnsxnxn committed:
style: 프론트 알림 페이지 수정
- 알림 페이지 로직 수정
- 알림 컨트롤러 수정

duswnsxnxn committed:
config: yml prod모드

duswnsxnxn committed:
config: yml수정

duswnsxnxn committed:
config: yml변경

duswnsxnxn committed:
docs: readme 작성

duswnsxnxn committed:
config: yml 수정

duswnsxnxn committed:
config: yml재설정

duswnsxnxn committed:
config: custom.site.baseUrl 도입하여 중복 제거

duswnsxnxn committed:
config: yml파일 baseUrl수정

duswnsxnxn committed:
refactor: 누락된 코드 수정

duswnsxnxn committed:
docs: records수정

Hjun1004 commented on src/main/java/com/ll/gramgram/boundedContext/notification/entity/Notification.java:
```diff
@@ -29,4 +33,8 @@ public class Notification extends BaseEntity {
+
+    public String getReadDateStr() {
+        return readDate.format(DateTimeFormatter.ofPattern("HH시:mm분, " + ChronoUnit.MINUTES.between(readDate, LocalDateTime.now()) + "분 전"));
+    }
```
notification/list.html에서 알림 부분에 getReadDateStr을 이용해서 시간을 표시하셨는데
알림 부분에서 나오는 시간은 알림이 온 시간인 createDate를 이용해서 표시하시는게 어떨까 싶어요!

duswnsxnxn replied:
아 그래야겠어요.생각해보니 제가 구현한 로직은 읽고 나서야 실행되는거라 의미가 없네요 ㅠㅠ 자세히 봐주셔서 감사합니다!!

Hjun1004 commented on src/main/resources/templates/usr/notification/list.html:
```diff
@@ -36,22 +36,22 @@ <h1 class="mb-4">
             </div>
 
-            <div class="card bg-base-100 shadow-xl">
+            <div class="card bg-base-100 shadow-xl" th:if="${notification.getTypeCode() == 'ModifyAttractiveType'}">
```
list.html에서 Like와 Modify에서 표시되는 부분이 다른건 내용 부분이기 때문에 전체에 IF를 거는것 보다 내용 밑에부분의 div에 IF를 걸어주시는게 코드도 줄어들고 중복되는 부분도 없앨수 있을것 같습니다!

duswnsxnxn replied:
와 그러네요.전체에다 걸 필요가 없었어요.진짜 감사합니다.많이 참고 될 것 같아요!!

Hjun1004 commented on src/main/java/com/ll/gramgram/boundedContext/notification/controller/NotificationController.java:
```diff
@@ -11,23 +14,34 @@
+            if (likeablePerson.isPresent()) {
+                LikeablePerson likeable = likeablePerson.get();
+                model.addAttribute("likeable", likeable);
+            }
```
model로 LikeablePerson을 넘겨주시는게 List에서 호감 사유 때문이신가요?
그런거라면 notification을 생성하실 때 newAttractiveTypeCode에 바로 넣어서 생성하시면 이 부분은 생략하실 수 있을것 같습니다!

duswnsxnxn replied:
알림 엔티티에 있는 oldAttractiveTypeCode, newAttractiveTypeCode 필드들은 수정할 때만 쓰는줄 알고 저렇게 했어요.다른 분들 하신거 보니까 말씀하신대로 저 필드들 이용해서 출력하셨더라고여. 리펙토링 작업에 참고 하려구여 조언 감사합니다.

Hjun1004 commented on records/3Week_김연준.md:
```diff
@@ -0,0 +1,47 @@
+- update를 하려는 엔티티가 빌더패턴으로 작성되있다면 어떻게 처리를 하는지 궁금합니다. 
+- 저는 빌더패턴을 잘 몰라 따로 update메소드를 해당 엔티티에 추가적으로 구현해주는 방식으로 처리했는데 이 방법으로 하는게 맞나요?
+- 테이블로부터 조인해서 가져올 때 jpa레파지토리의 쿼리 메소드 방식과 jpql을 이용한 방식 둘 중 뭐가 주로 쓰이는지 궁금합니다.
+
```
record가 필요한 부분만 정확하게 잘 적으셔서 읽기 편했고 이해가 너무 쉬웠습니다!
요점을 잘 짚으시는것 같아서 부러워요 ㅎㅎ

duswnsxnxn replied:
아이고 과찬이십니다~~
평가 좋은 블로그 글들을 참고해서 작성해봤는데 반응이 좋으시니 저도 좋네요 ㅎ

Hjun1004 commented on src/test/java/com/ll/gramgram/boundedContext/likeablePerson/service/LikeablePersonServiceTests.java:
```diff
@@ -263,4 +265,19 @@ void t008() throws Exception {
+        Assertions.assertThat(RsMsg).
+                isEqualTo(coolTime.format(DateTimeFormatter.ofPattern("HH시:mm분:ss초")) + "에 가능합니다."
+        );
+    }
```
테스트 케이스를 잘 활용하시는것 같아 부럽습니다.
저는 아직 테스트 케이스를 작성하는게 어려워서 활용을 못하는데 대단하신것 같아요!

duswnsxnxn replied:
저는 mockmvc 이용하는 부분에선 막히드라고여 ㅠ 테스트케이스 잘 짜는게 어렵네용 저도

Hjun1004 commented on src/main/java/com/ll/gramgram/boundedContext/notification/service/NotificationService.java:
```diff
@@ -1,19 +1,65 @@
+                .readDate(null)
+                .typeCode("Like")
+                .build();
+
```
notification을 생성할 때 저도 몰랐는데 null값이 들어가는 부분들은 생략해주어도 null이 들어간다고 합니다.
null이 들어갔다는것을 명시적으로 보여주기 위해서 남겨두셔도 좋을 것 같고
코드를 조금 줄이고자 한다면 생략하셔도 될것 같아요. 저도 null을 다 넣어주었는데 저는 빼는 쪽으로 리팩토링 하려고 합니다 ㅎㅎ

duswnsxnxn replied:
오 그런 성질이 있었군요 참고할게요 ㅎ

duswnsxnxn reviewed (commented):
꼼꼼하게 리뷰 해주셔서 감사합니다.
조언해주신 알림 페이지랑 알림 몇 분 전 부분
마지막으로 알림 엔티티의 AttractiveTypeCode필드 활용하는 부분까지 리펙토링 할 때 꼭 참고할게요!!
코드 봐주시느라 고생 많으셨습니다.

JunTaeINC commented on src/main/java/com/ll/gramgram/boundedContext/likeablePerson/entity/LikeablePerson.java:
```diff
@@ -40,7 +41,7 @@ public boolean isModifyUnlocked() {
     // 초 단위에서 올림 해주세요.
     public String getModifyUnlockDateRemainStrHuman() {
-        return "2시간 16분";
+        return modifyUnlockDate.format(DateTimeFormatter.ofPattern("HH시:mm분:ss초"));
```
아니.. 이런방법이..🤯

JunTaeINC reviewed (commented):
우와.. 연준님 엄청잘하셨네요.. 코드도 전체적으로 이해하기쉽게 잘 작셨네요;;
저에겐 선택미션은 태산이였네요... 코드 잘보고 잘배우고 갑니다 ㅎㅎ

duswnsxnxn committed:
refactor: 알림 페이지 불필요한 코드 수정

duswnsxnxn committed:
refactor: 리펙토링
- notification list 페이지 likeablePerson대신에 notifications에서 AttractiveType참조하도록 수정
- LikeablePersonService의 canCancel메소드에서 쿨타임 적용
- Notification엔티티의 readDate 로직 수정
